tests/phpunit/tests/query/cacheResults.php

```diff
@@ -0,0 +1,707 @@
+<?php
+
+/**
+ * @group query
+ */
+class Test_Query_CacheResults extends WP_UnitTestCase {
+	/**
+	 * Page IDs.
+	 *
+	 * @var int[]
+	 */
+	public static $pages;
+
+	/**
+	 * Post IDs.
+	 *
+	 * @var int[]
+	 */
+	public static $posts;
+
+	/**
+	 * Term ID.
+	 *
+	 * @var int
+	 */
+	public static $t1;
+
+	public function set_up() {
+		parent::set_up();
+		// Make some post objects.
+		self::$posts = self::factory()->post->create_many( 5 );
+		self::$pages = self::factory()->post->create_many( 5, array( 'post_type' => 'page' ) );
+
+		self::$t1 = self::factory()->term->create(
+			array(
+				'taxonomy' => 'category',
+				'slug'     => 'foo',
+				'name'     => 'Foo',
+			)
+		);
+
+		wp_set_post_terms( self::$posts[0], self::$t1, 'category' );
+		add_post_meta( self::$posts[0], 'color', '#000000' );
+	}
+
+	/**
+	 * @dataProvider data_query_args
+	 * @ticket 22176
+	 */
+	public function test_query_cache( $args ) {
+		$query1 = new WP_Query();
+		$posts1 = $query1->query( $args );
+
+		$queries_before = get_num_queries();
+		$query2         = new WP_Query();
+		$posts2         = $query2->query( $args );
+		$queries_after  = get_num_queries();
+
+		if ( isset( $args['fields'] ) ) {
+			if ( 'all' !== $args['fields'] ) {
+				$this->assertSameSets( $posts1, $posts2 );
+			}
+			if ( 'id=>parent' !== $args['fields'] ) {
+				$this->assertSame( $queries_after, $queries_before );
+			}
+		} else {
+			$this->assertSame( $queries_after, $queries_before );
+		}
+		$this->assertSame( $query1->found_posts, $query2->found_posts );
+		$this->assertSame( $query1->max_num_pages, $query2->max_num_pages );
+
+		if ( ! $query1->query_vars['no_found_rows'] ) {
+			wp_delete_post( self::$posts[0], true );
+			wp_delete_post( self::$pages[0], true );
+			$query3 = new WP_Query();
+			$query3->query( $args );
+
+			$this->assertNotSame( $query1->found_posts, $query3->found_posts );
+			$this->assertNotSame( $queries_after, get_num_queries() );
+		}
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[] Test parameters.
+	 */
+	public function data_query_args() {
```

Comment:
```suggestion
	public function data_query_cache() {
```

Convention is to use the same method name as the test but swap out the `test` prefix with `data`.